Panic during shutdown

After killing the agent with a ctrl-c:

```
level=info ts=2020-09-02T23:29:54.237414373Z caller=wal.go:423 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="WAL checkpoint complete" first=189 last=195 duration=41.297111278s
^Clevel=info ts=2020-09-02T23:34:54.52948527Z caller=signals.go:55 msg="=== received SIGINT/SIGTERM ===\n*** exiting"
level=info ts=2020-09-02T23:34:54.52987509Z caller=instance.go:297 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="stopping truncation loop..."
level=info ts=2020-09-02T23:34:54.529886854Z caller=instance.go:319 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="stopping scrape manager..."
level=info ts=2020-09-02T23:34:54.530004594Z caller=instance.go:526 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="discovery manager stopped"
level=info ts=2020-09-02T23:34:54.530108567Z caller=instance.go:529 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="stopping discovery manager..."
level=info ts=2020-09-02T23:34:54.872412398Z caller=instance.go:332 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="closing storage..."
level=info ts=2020-09-02T23:34:54.872577813Z caller=instance.go:313 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="scrape manager stopped"
ts=2020-09-02T23:34:54.872700988Z caller=dedupe.go:112 agent=prometheus instance=5df40780494f25472649ea5095966285 component=remote level=info remote_name=cortex url=http://localhost:1234/receive msg="Stopping remote storage..."
ts=2020-09-02T23:34:54.90359832Z caller=dedupe.go:112 agent=prometheus instance=5df40780494f25472649ea5095966285 component=remote level=info remote_name=cortex url=http://localhost:1234/receive msg="WAL watcher stopped" queue=cortex
ts=2020-09-02T23:34:54.903633143Z caller=dedupe.go:112 agent=prometheus instance=5df40780494f25472649ea5095966285 component=remote level=info remote_name=cortex url=http://localhost:1234/receive msg="Stoping metadata watcher..."
ts=2020-09-02T23:34:54.903674878Z caller=dedupe.go:112 agent=prometheus instance=5df40780494f25472649ea5095966285 component=remote level=info remote_name=cortex url=http://localhost:1234/receive msg="Scraped metadata watcher stopped"
level=info ts=2020-09-02T23:34:55.870339333Z caller=wal.go:358 agent=prometheus instance=5df40780494f25472649ea5095966285 msg="series GC completed" duration=1.358675992s
panic: send on closed channel

goroutine 750 [running]:
github.com/prometheus/prometheus/tsdb/wal.(*WAL).nextSegment(0xc000392870, 0x33, 0xc2ee8bd800)
        /home/runner/work/agent/agent/vendor/github.com/prometheus/prometheus/tsdb/wal/wal.go:481 +0xf6
github.com/prometheus/prometheus/tsdb/wal.(*WAL).NextSegment(0xc000392870, 0x0, 0x0)
        /home/runner/work/agent/agent/vendor/github.com/prometheus/prometheus/tsdb/wal/wal.go:460 +0x78
github.com/grafana/agent/pkg/prom/wal.(*Storage).Truncate(0xc000c784e0, 0x174512c5530, 0x4, 0x4)
        /home/runner/work/agent/agent/pkg/prom/wal/wal.go:367 +0x24f
github.com/grafana/agent/pkg/prom/instance.(*Instance).truncateLoop(0xc0003b0c60, 0x2caf5c0, 0xc0a3a34ac0, 0x2ccffe0, 0xc000c784e0, 0xc00078e060)
        /home/runner/work/agent/agent/pkg/prom/instance/instance.go:581 +0x3b8
github.com/grafana/agent/pkg/prom/instance.(*Instance).Run.func1(0x0, 0x0)
        /home/runner/work/agent/agent/pkg/prom/instance/instance.go:292 +0x60
github.com/oklog/run.(*Group).Run.func1(0xc0a3ac8600, 0xc0a3ddb500, 0xc0a37b9d40)
        /home/runner/work/agent/agent/vendor/github.com/oklog/run/group.go:38 +0x27
created by github.com/oklog/run.(*Group).Run
        /home/runner/work/agent/agent/vendor/github.com/oklog/run/group.go:37 +0xbb
```

```
$ uname -a
Linux ip-10-1-5-195 4.15.0-1035-aws #37-Ubuntu SMP Mon Mar 18 16:15:14 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux
```

```
$ ./agent-linux-amd64 --version
agent, version v0.5.0 (branch: HEAD, revision: 86f7b71)
  build user:       runner@fv-az151
  build date:       2020-08-12T14:40:24Z
  go version:       go1.14.4
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -61,6 +61,9 @@ this platform. FreeBSD builds will return in a future release.
   remote_write names will be generated based on the first 6 characters of the
   group hash and the first six characters of the remote_write hash. (@rfratto)
 
+- [BUGFIX] Fix a panic that may occur during shutdown if the WAL is closed in
+  the middle of the WAL being truncated. (@rfratto)
+
 - [DEPRECATION] `use_hostname_label` is now supplanted by
   `replace_instance_label`. `use_hostname_label` will be removed in a future
   version. (@rfratto)

--- a/pkg/prom/wal/wal.go
+++ b/pkg/prom/wal/wal.go
@@ -19,6 +19,10 @@ import (
 	"github.com/prometheus/prometheus/tsdb/wal"
 )
 
+var (
+	ErrWALClosed = fmt.Errorf("WAL storage closed")
+)
+
 type storageMetrics struct {
 	r prometheus.Registerer
 
@@ -90,6 +94,12 @@ type Storage struct {
 	storage.Queryable
 	storage.ChunkQueryable
 
+	// Operations against the WAL must be protected by a mutex so it doesn't get closed
+	// in the middle of an operation. If the WAL is closed, operations that change the WAL
+	// must fail.
+	walMtx    sync.Mutex
+	walClosed chan bool
+
 	path   string
 	wal    *wal.WAL
 	logger log.Logger
@@ -115,6 +125,8 @@ func NewStorage(logger log.Logger, registerer prometheus.Registerer, path string
 	}
 
 	storage := &Storage{
+		walClosed: make(chan bool),
+
 		path:    path,
 		wal:     w,
 		logger:  logger,
@@ -151,6 +163,9 @@ func NewStorage(logger log.Logger, registerer prometheus.Registerer, path string
 }
 
 func (w *Storage) replayWAL() error {
+	w.walMtx.Lock()
+	defer w.walMtx.Unlock()
+
 	level.Info(w.logger).Log("msg", "replaying WAL, this may take a while", "dir", w.wal.Dir())
 	dir, startFrom, err := wal.LastCheckpoint(w.wal.Dir())
 	if err != nil && err != record.ErrNotFound {
@@ -352,6 +367,13 @@ func (*Storage) StartTime() (int64, error) {
 // Truncate removes all data from the WAL prior to the timestamp specified by
 // mint.
 func (w *Storage) Truncate(mint int64) error {
+	w.walMtx.Lock()
+	defer w.walMtx.Unlock()
+
+	if w.isWALClosed() {
+		return ErrWALClosed
+	}
+
 	start := time.Now()
 
 	// Garbage collect series that haven't received an update since mint.
@@ -511,10 +533,27 @@ func (w *Storage) WriteStalenessMarkers(remoteTsFunc func() int64) error {
 
 // Close closes the storage and all its underlying resources.
 func (w *Storage) Close() error {
+	w.walMtx.Lock()
+	defer w.walMtx.Unlock()
+
+	if w.isWALClosed() {
+		return fmt.Errorf("already closed")
+	}
+	close(w.walClosed)
+
 	if w.metrics != nil {
 		w.metrics.Unregister()
 	}
 	return w.wal.Close()
+}
+
+func (w *Storage) isWALClosed() bool {
+	select {
+	case <-w.walClosed:
+		return true
+	default:
+		return false
+	}
 }
 
 type appender struct {
@@ -578,6 +617,13 @@ func (a *appender) AddFast(ref uint64, t int64, v float64) error {
 
 // Commit submits the collected samples and purges the batch.
 func (a *appender) Commit() error {
+	a.w.walMtx.Lock()
+	defer a.w.walMtx.Unlock()
+
+	if a.w.isWALClosed() {
+		return ErrWALClosed
+	}
+
 	var encoder record.Encoder
 	buf := a.w.bufPool.Get().([]byte)
 

--- a/pkg/prom/wal/wal.go
+++ b/pkg/prom/wal/wal.go
@@ -166,6 +166,10 @@ func (w *Storage) replayWAL() error {
 	w.walMtx.Lock()
 	defer w.walMtx.Unlock()
 
+	if w.isWALClosed() {
+		return ErrWALClosed
+	}
+
 	level.Info(w.logger).Log("msg", "replaying WAL, this may take a while", "dir", w.wal.Dir())
 	dir, startFrom, err := wal.LastCheckpoint(w.wal.Dir())
 	if err != nil && err != record.ErrNotFound {

--- a/pkg/prom/wal/wal.go
+++ b/pkg/prom/wal/wal.go
@@ -94,11 +94,12 @@ type Storage struct {
 	storage.Queryable
 	storage.ChunkQueryable
 
-	// Operations against the WAL must be protected by a mutex so it doesn't get closed
-	// in the middle of an operation. If the WAL is closed, operations that change the WAL
-	// must fail.
-	walMtx    sync.Mutex
-	walClosed chan bool
+	// Operations against the WAL must be protected by a mutex so it doesn't get
+	// closed in the middle of an operation. Other operations are concurrency-safe, so we
+	// use a RWMutex to allow multiple usages of the WAL at once. If the WAL is closed, all
+	// operations that change the WAL must fail.
+	walMtx    sync.RWMutex
+	walClosed bool
 
 	path   string
 	wal    *wal.WAL
@@ -125,8 +126,6 @@ func NewStorage(logger log.Logger, registerer prometheus.Registerer, path string
 	}
 
 	storage := &Storage{
-		walClosed: make(chan bool),
-
 		path:    path,
 		wal:     w,
 		logger:  logger,
@@ -163,10 +162,10 @@ func NewStorage(logger log.Logger, registerer prometheus.Registerer, path string
 }
 
 func (w *Storage) replayWAL() error {
-	w.walMtx.Lock()
-	defer w.walMtx.Unlock()
+	w.walMtx.RLock()
+	defer w.walMtx.RUnlock()
 
-	if w.isWALClosed() {
+	if w.walClosed {
 		return ErrWALClosed
 	}
 
@@ -371,10 +370,10 @@ func (*Storage) StartTime() (int64, error) {
 // Truncate removes all data from the WAL prior to the timestamp specified by
 // mint.
 func (w *Storage) Truncate(mint int64) error {
-	w.walMtx.Lock()
-	defer w.walMtx.Unlock()
+	w.walMtx.RLock()
+	defer w.walMtx.RUnlock()
 
-	if w.isWALClosed() {
+	if w.walClosed {
 		return ErrWALClosed
 	}
 
@@ -540,24 +539,15 @@ func (w *Storage) Close() error {
 	w.walMtx.Lock()
 	defer w.walMtx.Unlock()
 
-	if w.isWALClosed() {
+	if w.walClosed {
 		return fmt.Errorf("already closed")
 	}
-	close(w.walClosed)
+	w.walClosed = true
 
 	if w.metrics != nil {
 		w.metrics.Unregister()
 	}
 	return w.wal.Close()
-}
-
-func (w *Storage) isWALClosed() bool {
-	select {
-	case <-w.walClosed:
-		return true
-	default:
-		return false
-	}
 }
 
 type appender struct {
@@ -621,10 +611,10 @@ func (a *appender) AddFast(ref uint64, t int64, v float64) error {
 
 // Commit submits the collected samples and purges the batch.
 func (a *appender) Commit() error {
-	a.w.walMtx.Lock()
-	defer a.w.walMtx.Unlock()
+	a.w.walMtx.RLock()
+	defer a.w.walMtx.RUnlock()
 
-	if a.w.isWALClosed() {
+	if a.w.walClosed {
 		return ErrWALClosed
 	}
 

--- a/pkg/prom/wal/wal_test.go
+++ b/pkg/prom/wal/wal_test.go
@@ -239,6 +239,18 @@ func TestStorage_WriteStalenessMarkers(t *testing.T) {
 	}
 }
 
+func TestStoraeg_TruncateAfterClose(t *testing.T) {
+	walDir, err := ioutil.TempDir(os.TempDir(), "wal")
+	require.NoError(t, err)
+	defer os.RemoveAll(walDir)
+
+	s, err := NewStorage(log.NewNopLogger(), nil, walDir)
+	require.NoError(t, err)
+
+	require.NoError(t, s.Close())
+	require.Error(t, ErrWALClosed, s.Truncate(0))
+}
+
 type sample struct {
 	ts  int64
 	val float64

